presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/LogicalCteOptimizer.java

```diff
@@ -0,0 +1,197 @@
+
+public class LogicalCteOptimizer
+        implements PlanOptimizer
+{
+    private final Metadata metadata;
+
+    public LogicalCteOptimizer(Metadata metadata)
+    {
+        this.metadata = metadata;
+    }
+
+    @Override
+    public PlanOptimizerResult optimize(PlanNode plan, Session session, TypeProvider types, VariableAllocator variableAllocator, PlanNodeIdAllocator idAllocator, WarningCollector warningCollector)
+    {
+        if (!isMaterializeAllCtes(session)) {
+            return PlanOptimizerResult.optimizerResult(plan, false);
+        }
+        PlanNode rewrittenPlan = new CteEnumerator(session.getCteInformationCollector(), idAllocator, variableAllocator).transformPersistentCtes(plan);
+        return PlanOptimizerResult.optimizerResult(rewrittenPlan, !rewrittenPlan.equals(plan));
+    }
+
+    public class CteEnumerator
+    {
+        private final CTEInformationCollector cteInformationCollector;
+        PlanNodeIdAllocator planNodeIdAllocator;
+        VariableAllocator variableAllocator;
+
+        public CteEnumerator(CTEInformationCollector cteInformationCollector, PlanNodeIdAllocator planNodeIdAllocator, VariableAllocator variableAllocator)
+        {
+            this.cteInformationCollector = cteInformationCollector;
+            this.planNodeIdAllocator = planNodeIdAllocator;
+            this.variableAllocator = variableAllocator;
+        }
+
+        public PlanNode transformPersistentCtes(PlanNode root)
+        {
+            checkArgument(root.getSources().size() == 1, "expected newChildren to contain 1 node");
+            CteTransformerContext context = new CteTransformerContext();
+            PlanNode transformedCte = SimplePlanRewriter.rewriteWith(new CteConsumerTransformer(planNodeIdAllocator, variableAllocator),
+                    root, context);
+            List<PlanNode> topologicalOrderedList = context.getTopologicalOrdering();
+            if (topologicalOrderedList.isEmpty()) {
+                return root;
+            }
+            SequenceNode sequenceNode = new SequenceNode(root.getSourceLocation(), planNodeIdAllocator.getNextId(), topologicalOrderedList,
+                    transformedCte.getSources().get(0));
+            return root.replaceChildren(Arrays.asList(sequenceNode));
+        }
+    }
+
+    public class CteConsumerTransformer
+            extends SimplePlanRewriter<CteTransformerContext>
+    {
+        private final PlanNodeIdAllocator idAllocator;
+
+        private final VariableAllocator variableAllocator;
+
+        public CteConsumerTransformer(PlanNodeIdAllocator idAllocator, VariableAllocator variableAllocator)
+        {
+            this.idAllocator = idAllocator;
+            this.variableAllocator = variableAllocator;
+        }
+
+        @Override
+        public PlanNode visitCteReference(CteReferenceNode node, RewriteContext<CteTransformerContext> context)
+        {
+            context.get().addDependency(node.getCteName());
```

Comment:
Is the name returned by `node.getCteName()` unique to identify the CTE here? 
For example, will it be able to differentiate the two `t2` in the query here?
`with t1 as (with t2 as (select * from (values 1, 2)t(k)) select k+1 as k from t2), t2 as (select k+1 as k from t1) select k+1 from t2;`

Reply:
Hmm interesting, I am not sure if this is allowed in Sql Standard.

Reply:
I think nested ctes are not allowed in SQL standard, but for us it makes sense to add a test case/fix for this

Reply:
Not sure about SQL standard (maybe @kaikalur know it), but it's a valid SQL which can run in Presto

Reply:
please also check the case with view names: I believe a query can reference a view and a cte with the same name and there are some rules around how name resolution is done